### DISCUSSION TITLE
Cameras within references

### DISCUSSION THREAD

**Discussion by gracekumagai** [Autodesk/maya-usd#2008] [Q&A] [answered]
We are running into an issue where we have 100s of cameras listed in the Viewport's Panels -> Perspective menu when loading a full set. It can take a long time to open the list of cameras and it can be challenging to find the camera you are looking for.

Our sets are constructed by referencing in models and a model can contain several projection cameras. We were wondering if there has been any discussion on what cameras should/should not be listed? Or, if we may want a setting to control what cameras are listed in the drop down?

Also, if it would be preferred to handle this on a per-studio basis, is there a way we can hook into this menu and customize the UsdCameras that are listed?

Thank you!

**Comment by williamkrick** [comment]
Looking at the mel to generate this menu (buildPerspLookthruMenu.mel in runTime\scripts\others) there is currently no way to customize or filter the list. The listCameras command returns all the perspective cameras in the scene (native Maya and Ufe) and builds a menu item for each one.

@seando-adsk I'm not familiar enough with the UI to know if/how we can make it configurable. Do you have any ideas?

**Comment by williamkrick** [comment]
Chatting internally we came up with an idea. If you replace the UsdCameraHandler with your own camera handler (by inserting your handler before ours https://github.com/Autodesk/maya-usd/blob/91ced343baa0613633eae6f4d47ed396c433defd/lib/mayaUsd/ufe/Global.cpp#L154). In your custom handler you would return a nullptr scene item if you don't want that camera path to show up as a camera in Maya.

That will prevent Maya from finding the camera in buildPrespLookthruMenu.mel.

**Reply by gracekumagai** [reply]
Oh awesome! Thank you for looking into it. We'll give that a shot!

**Comment by ppt-adsk** [comment, marked as answer]
Hi Grace, hopefully when you try it out it'll suit your purpose.  It's a bit of a blunt instrument, in that the CameraHandler isn't passed any context to decide whether or not to create the Ufe::Camera interface, but the idea is that relying on the argument Ufe::SceneItem is enough.  It might look like this:
```
class PxrCameraHandler : public Ufe::CameraHandler {

PxrCameraHandler(const Ufe::CameraHandler::Ptr& nextHandler) : _nextHandler(nextHandler) {}

Camera::Ptr camera(const SceneItem::Ptr& item) const override {
    return dontWannaSee(item) ? nullptr : _nextHandler->camera(item);
}
};
```
Hope this helps.  Cheers.

**Comment by gracekumagai** [comment]
Thank you for your help! We were able to make an internal patch to filter out the cameras. 

It is still a bit slow to open the menu on heavy sets (since it has to go through the scene items looking for cameras). However, once it is open, it is easier to find the camera we are looking for - thank you!

**Reply by kxl-adsk** [reply]
Performance impact is probably coming from our traversal over the entire scene to discover cameras. What would be the most performant way to do this in USD?

**Reply by gracekumagai** [reply]
From chatting internally, they suggested using `UsdPrim::IsA<UsdGeomCamera>()` instead of constructing UsdGeomCamera and checking if it's valid.

I took a look around and I was wondering if the traversal code is open sourced? For our workflow, we would be able to stop the traversal at the root of the referenced models (since we don't want to display cameras within models).

**Reply by kxl-adsk** [reply]
Thank you, @gracekumagai. Traversal currently is done from Maya via ufe::hierarchy interface. We would need a dedicated method to discover items/paths of a given type to leverage any potential optimizations for this task in the particular data model.
